### PR TITLE
Convert more of tikv crate to engine traits

### PR DESCRIPTION
**Pull request by brson**
### What problem does this PR solve?

This converts more of the tikv crate to use engine traits instead of rocks directly. This is pretty much all of the easy stuff in the tikv crate. Everything else is going to take careful thought.

Part of https://github.com/tikv/tikv/issues/6402

Problem Summary:

### Release note 

- No release note.

### CODE REVIEW

File: src/server/raft_client.rs [hunk 1]
```diff
@@ -288,14 +289,19 @@ impl Buffer for MessageBuffer {
...
     region_id: u64,
     to_peer_id: u64,
     to_store_id: u64,
+    _engine: PhantomData<E>,
 }
 
-impl<T: RaftStoreRouter<RocksEngine> + 'static> SnapshotReporter<T> {
+impl<T, E> SnapshotReporter<T, E>
+where
+    T: RaftStoreRouter<E> + 'static,
```
Comment (Little-Wallace) [inline src/server/raft_client.rs:302]: Could it be `RaftStoreRouter<impl KvEngine >`?

Comment (brson) [reply src/server/raft_client.rs:302]: I looked into it and impl trait is not valid in this position.

**Comment by brson** [comment]
This Pr is updated and I've responded to the review question. cc @Little-Wallace @Connor1996 @hicqu

**Review by Connor1996 — COMMENTED** [review]
Can we make `CompactEvent` be `Box<dyn>` to remove the dependency on engine_trait for RaftRouter like https://github.com/tikv/tikv/pull/8983?
Furthermore, we can try to remove `CompactEvent` from `engine_trait` though it needs an elaborate redesign. We can leave it for future work.

**Comment by Little-Wallace** [comment]
I agree. I think we do not need to add more type parameter...

**Comment by brson** [comment]
@Little-Wallace can you get https://github.com/tikv/tikv/pull/8983 ready to merge?

It seems like an improvement and a prerequisite for anything in this patch.

**Comment by brson** [comment]
Blocked on https://github.com/tikv/tikv/pull/8983

**Comment by brson** [comment]
This patch is hopelessly bitrotted. Closing.

**Pull request closed without merging**